File: public/javascripts/clientGame.js
define(["jquery", "ui/homeScreen", "ui/preGame", "ui/inGame", "playerEngine", "networking/homeScreen", "networking/preGame", "networking/inGame"],
function ($, _homeScreen, _preGame, _inGame, _playerEngine, _homeScreenComm, _preGameComm, _inGameComm) {
    class Game {
        constructor(container) {
            this.mainContainer = $(container);
            this.userInterface = new _homeScreen.HomeScreenUserInterface(
                Game.createHomeScreenComm(),
                this,
                this.mainContainer
            )
        }

        initiatePreGame(response, isGameCreator) {
            this.userInterface.destroy();

            this.userInterface = new _preGame.PreGameUserInterface(
                Game.createPreGameComm(response),
                response.simpleId, isGameCreator,
                this,
                this.mainContainer
            );
        }

        initiateGame(response) {
            this.userInterface.destroy();
            var playerEngine = new _playerEngine.PlayerEngine(Game.createInGameComm(response), response.factions);
            this.userInterface = new _inGame.InGameUserInterface(
                playerEngine,
                this.mainContainer
            )
        }

        // create networking services
        static createHomeScreenComm() {
            return new _homeScreenComm.HomeScreenCommunication();
        }

        static createPreGameComm(response) {
            return new _preGameComm.PreGameCommunication(response.gameId, response.faction.id);
        }

        static createInGameComm(response) {
            return new _inGameComm.InGameCommunication(response.gameId, response.factions[0].id, response.systems);
        }
    }

    return {
        Game: Game
    };
});

File: src/client/ClientGame.scala
package client

import client.ui.{HomeScreenUserInterface, InGameUserInterface, PreGameUserInterface, UserInterface}
import networking.dto.{BeginResponse, PreGameResponse}
import networking.{HomeScreenCommunication, InGameCommunication, PreGameCommunication}
import org.newdawn.slick.{BasicGame, GameContainer, Graphics}

/**
  * Created by dan-s on 24/05/2016.
  */

trait ClientGame { self: BasicGame =>
  private var userInterface: UserInterface = _
  override def init(gameContainer: GameContainer) = {
    userInterface = new HomeScreenUserInterface(this, createHomeScreenComm(), gameContainer)
  }

  override def render(gameContainer: GameContainer, graphics: Graphics): Unit = {
    if (userInterface != null) {
      userInterface.render(gameContainer, graphics)
    } else {
      throw new Exception("User interface failed to be initialized")
    }
  }
  def initiatePreGame(response: PreGameResponse, isOwner: Boolean): Unit = {
    userInterface = new PreGameUserInterface(
      createPreGameComm(response),
      response.simpleId,
      isOwner,
      this
    )
  }
  def initiateGame(response: BeginResponse): Unit = {
    userInterface = new InGameUserInterface(
      new PlayerEngine(
        createInGameComm(response)
      )
    )
  }

  protected def createHomeScreenComm(): HomeScreenCommunication
  protected def createPreGameComm(response: PreGameResponse): PreGameCommunication
  protected def createInGameComm(response: BeginResponse): InGameCommunication
}
